sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,203 @@
  */
 class SynchronousMessageSubscriber extends BaseSubscriber<ServiceBusReceivedMessageContext> {
     private final ClientLogger logger = new ClientLogger(SynchronousMessageSubscriber.class);
-    private final Timer timer = new Timer();
     private final AtomicBoolean isDisposed = new AtomicBoolean();
-    private final SynchronousReceiveWork work;
+    private final long prefetch;
+    private final AtomicInteger wip = new AtomicInteger();
+    private final Queue<SynchronousReceiveWork> workQueue = new ConcurrentLinkedQueue<>();
+    private final Queue<ServiceBusReceivedMessageContext> bufferMessages = new ConcurrentLinkedQueue<>();
+    private final SynchronousReceiveWork initialWork;
+    private final AtomicLong remaining = new AtomicLong();
 
     private volatile Subscription subscription;
 
-    SynchronousMessageSubscriber(SynchronousReceiveWork work) {
-        this.work = Objects.requireNonNull(work, "'work' cannot be null.");
+    private SynchronousReceiveWork currentWork;
+    private Disposable timeoutOperation;
+
+    SynchronousMessageSubscriber(long prefetch, SynchronousReceiveWork initialWork) {
+        this.prefetch = prefetch;
+        this.initialWork = initialWork;
     }
 
     /**
      * On an initial subscription, will take the first work item, and request that amount of work for it.
-     *
      * @param subscription Subscription for upstream.
      */
     @Override
     protected void hookOnSubscribe(Subscription subscription) {
-        if (this.subscription == null) {
-            this.subscription = subscription;
-        }
+        this.subscription = subscription;
 
-        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", work.getId(), work.getNumberOfEvents(),
-            work.getTimeout());
-
-        subscription.request(work.getNumberOfEvents());
+        logger.info("[{}] onSubscribe Pending: {}, Scheduling receive timeout task '{}'.", initialWork.getId(),
+            initialWork.getNumberOfEvents(), initialWork.getTimeout());
 
-        timer.schedule(new ReceiveTimeoutTask(work.getId(), this::dispose), work.getTimeout().toMillis());
+        // This will trigger subscription.request(N) and queue up the work
+        queueWork(initialWork);
     }
 
     /**
      * Publishes the event to the current {@link SynchronousReceiveWork}. If that work item is complete, will dispose of
      * the subscriber.
-     *
-     * @param value Event to publish.
+     * @param message Event to publish.
      */
     @Override
-    protected void hookOnNext(ServiceBusReceivedMessageContext value) {
-        work.next(value);
+    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
+        if (currentWork == null) {
+            //Boundary condition(timeout cases), buffer the received message for future requests.
+            bufferMessages.add(message);
```

Comment:
Example: When there is only one receive(2) request and it time out. But we got two messages after that. During this time we also got request for another   receive(2) . In this case we can use this buffer to send the messages and do not need to ask upstream.